async-iterator.md

```diff
@@ -0,0 +1,904 @@
+- if you reach an `await`, you get a pending task and the code continues to execute in the background.
+
+As the code continues to execute in the background:
+- if it reaches a `yield return`, the background execution completes and the task we've previously handed to the caller is fulfilled with `true`,
+- if it reaches the end of the method, that task is fulfilled with `false`,
+- if it reaches an `await`, execution continues in the background.
+
+Note that the method is moved forward either by the caller or by background execution, but never both. Calling `MoveNextAsync()` before the task from the previous call complets yield unspecified results.
+
+
+### Exception handling (optional)
+
+When the caller is moving the method forward and an exception is thrown (without being caught by user code in the async-iterator method), `MoveNextAsync()` will return a task as normal.
+
+That task is complete (rather than pending). When the caller tries to access its result the task will throw the exception it holds.
+
+When background execution is moving the method forward and an exception is thrown without handling, the exception is similarly caught and passed onto the caller via the task.
+
+
+### Parallels with iterator and async control flow (optional)
+
+The execution of iterator methods shifts the control between the caller and the method:
+- the caller repeatedly calls `MoveNext()` (which executes the method),
+- the method yields control back to the caller (when reaching a `yield return`).
+
+The execution of async methods starts by the caller giving control to the method and background execution.
+From there on, the control shifts between the method and background execution:
+- the background execution resumes the method after a long-running task completes,
+- the method yields control back to the background execution (when reaching an `await`).
+Only when the method completes with a `return` is the control fully returned to the caller.
+
+The execution of async-iterator methods mixes both of those patterns, with the control shifting between caller (repeatedly calls `MoveNextAsync()`), method (gives control to the caller when reaching a `yield return` and to the background execution when reaching an `await`) and background execution (resumes the method after an `await`).
+
+
+## Lowering
+
+I'll describe the process of **lowering** async-iterator methods, that is replacing awaits and yields with simpler primitives.
+
+I'll start with many techniques already used for async methods (await suspensions, extracting locals to fields, spilling awaits, dispatch blocks).
+Then I'll explain `yield return` suspensions, disposal and yield breaks.
+
+
+### `await`
+
+`await expr` in async-iterator methods is lowered as:
+```C#
+... code before `await expr` ...
+
+{
+    var awaiterTemp = <expr>.GetAwaiter();
+    if(!awaiterTemp.IsCompleted)
+    {
+        /* set state to N */
+        /* save awaiter temp */
+        /* kick off background execution */
```

Comment:
i wouldn't mind explanations of these.  if those explanations come later, can the comment mention that fact?  For me, i like having the pseudo-code, knowing that each bit will be expanded on more fully.  Thanks!